File: Day<23>/Solution.cpp
#include <bits/stdc++.h> 
using namespace std; 


 // } Driver Code Ends
//User function Template for C++

class Solution{
    struct Info{
    	int distance;
    	int houseno;
    
    	Info(int x,int y)
    	{
    		distance = x;
    		houseno = y;
    	}
    };
    
    struct Compare{
    	bool operator() (Info const& p1,Info const& p2 )
    	{
    		if (p1.distance == p2.distance)	
    			return p1.houseno < p2.houseno;
    		else
    		    return p1.distance < p2.distance;
    	}
    };
    
    public:
    vector<int> Kclosest(vector<int>arr, int n, int x, int k) 
    { 
    	vector<int> result;
    	priority_queue <Info, vector<Info>, Compare> pq; 
    
    	for (int i = 0; i < k; i++) 
    	{
    		Info obj(abs(arr[i] - x) , arr[i] );
    		pq.push(obj); 
    	}
    
    	for (int i = k; i < n; i++) 
    	{ 
    		int diff = abs(arr[i] - x); 
    		if (diff > pq.top().distance) 
    			continue; 
    		
    		if (diff == pq.top().distance && arr[i] > pq.top().houseno) 
    			continue; 
    		
    		pq.pop();
    		Info obj(abs(arr[i] - x) , arr[i] );
    		pq.push(obj); 
    	} 
    	while (pq.empty() == false) { 
    		result.push_back(pq.top().houseno);
    		pq.pop();
    	}
    	sort(result.begin(),result.end());
    	return result;
    }      
};

// { Driver Code Starts.

int main() 
{ 
	int t;
	cin>>t;
	while(t--)
	{
		int n,x,k;
		cin>>n>>x>>k; 
		vector <int> array(n);
		for (int i = 0; i < n; ++i)
			cin>>array[i];
        
        Solution obj;
		vector <int> result = obj.Kclosest(array, n, x, k); 
		for (int i = 0; i < result.size(); ++i)
		{
			cout<<result[i]<<" ";
		}
		cout<<"\n";
	}

	return 0; 
}

File: Day<23>/Solution.java
import java.io.*;
import java.util.*; 
import java.lang.*;

class Info
{
	int distance;
	int houseno;
	Info(int x,int y)
	{
		distance = x;
		houseno = y;
	}
}

class Compare implements Comparator<Info>
{ 
	public int compare (Info p1,Info p2)
	{
		if (p1.distance == p2.distance)
		{
			if (p1.houseno < p2.houseno)
				return +1;
			if (p1.houseno > p2.houseno)
				return -1;
			return 0;
		}
		else
		{
		   	if (p1.distance < p2.distance)
				return +1;
			if (p1.distance > p2.distance)
				return -1;
			return 0;
		}
	}
}

class Solution
{
	public  ArrayList<Integer> Kclosest(int arr[], int n, int x, int k)
	{
		ArrayList<Integer> result= new ArrayList<Integer>();
		PriorityQueue<Info> pq = new PriorityQueue<Info>(k, new Compare());

		for (int i = 0; i < k; i++)
		{
			Info obj = new Info(Math.abs(arr[i] - x) , arr[i]);
			pq.add(obj);
		}

		for (int i = k; i < n; i++)
		{
			int diff = Math.abs(arr[i] - x);
			if (pq.peek().distance < diff)
				continue;

			if (diff == pq.peek().distance && pq.peek().houseno < arr[i])
				continue;

			pq.remove();
			Info obj = new Info(Math.abs(arr[i] - x) , arr[i]);
			pq.add(obj);
		}
		while (0 < pq.size())
		{
			result.add(pq.peek().houseno);
			pq.remove();
		}
		Collections.sort(result);
		return result;
	}
}


class GfG
{
    public static void main(String args[]) throws IOException
	{
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t > 0)
		{
        	int n = sc.nextInt();
        	int x = sc.nextInt();
        	int k = sc.nextInt();
        	int arr[] = new int[n];
        	for(int i=0; i<n; i++)
        	{
        		arr[i] = sc.nextInt();
        	}
            Solution ob = new Solution();
            ArrayList<Integer> ans = ob.Kclosest(arr,n,x,k);
            for (int i=0; i<ans.size(); i++)
        	    System.out.print(ans.get(i)+" ");
            System.out.println();
            t--;
        }
    }
}
